File: mobile/native/android/apps/ensu/app-ui/src/main/java/io/ente/ensu/whatsnew/WhatsNewService.kt
package io.ente.ensu.whatsnew

import android.content.Context

data class PendingWhatsNew(
    val version: Int,
    val entries: List<WhatsNewEntry>
)

class WhatsNewService(context: Context) {
    private val preferences = context.getSharedPreferences(
        PREFERENCES_NAME,
        Context.MODE_PRIVATE
    )

    fun getPendingWhatsNew(): PendingWhatsNew? {
        if (!preferences.contains(KEY_SEEN_VERSION)) {
            markSeen()
            return null
        }

        val seenVersion = preferences.getInt(KEY_SEEN_VERSION, WhatsNewContent.VERSION)
        if (seenVersion >= WhatsNewContent.VERSION) return null

        if (WhatsNewContent.entries.isEmpty()) {
            markSeen()
            return null
        }

        return PendingWhatsNew(
            version = WhatsNewContent.VERSION,
            entries = WhatsNewContent.entries
        )
    }

    fun markSeen() {
        preferences.edit()
            .putInt(KEY_SEEN_VERSION, WhatsNewContent.VERSION)
            .apply()
    }

    private companion object {
        const val PREFERENCES_NAME = "ensu_whats_new"
        const val KEY_SEEN_VERSION = "seen_version"
    }
}

File: mobile/native/apple/apps/ensu/Ensu/Services/WhatsNewService.swift
import Foundation

struct PendingWhatsNew: Identifiable {
    let version: Int
    let entries: [WhatsNewEntry]

    var id: Int { version }
}

final class WhatsNewService {
    static let shared = WhatsNewService()

    private let defaults: UserDefaults
    private let seenVersionKey = "ensu.whatsNew.seenVersion"

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    func pendingWhatsNew() -> PendingWhatsNew? {
        if defaults.object(forKey: seenVersionKey) == nil {
            markSeen()
            return nil
        }

        let seenVersion = defaults.integer(forKey: seenVersionKey)
        if seenVersion >= WhatsNewContent.version {
            return nil
        }

        if WhatsNewContent.entries.isEmpty {
            markSeen()
            return nil
        }

        return PendingWhatsNew(
            version: WhatsNewContent.version,
            entries: WhatsNewContent.entries
        )
    }

    func markSeen() {
        defaults.set(WhatsNewContent.version, forKey: seenVersionKey)
    }
}
